TaskManager.java

```diff
@@ -0,0 +1,70 @@
+@Component
+public class TaskManager {
+    private EntityManager entityManager;
+    private static TaskManager instance;
+    private LinkedList<Task> queue = new LinkedList<>();
+    private HashSet<Task> running = new HashSet<>();
+    private HashSet<Task> blocked = new HashSet<>();
+    private Thread checker = new Thread(() -> {
+        while (true) {
+            checkRun();
+        }
+    });
+
+    public TaskManager() {
+        instance = this;
+        checker.run();
+    }
+
+    public static TaskManager getInstance() {
+        return instance;
+    }
+
+    @Transactional
+    public void runTask(Task task) {
+        if (Collections.disjoint(task.getDependsOn(), running)) {
+            running.add(task);
+            task.setStatus(TaskStatus.Working);
+            new Thread(() -> task.runTask(entityManager)).run();
+            entityManager.persist(new TaskHistoryRecord("task " + task.getName() + " " + task.getSchedule() + " " +
+                    " started " + System.currentTimeMillis()));
+        } else {
+            blocked.add(task);
+            task.setStatus(TaskStatus.Blocked);
+            entityManager.persist(new TaskHistoryRecord("task " + task.getName() + " " + task.getSchedule() + " " +
+                    " blocked " + System.currentTimeMillis()));
+        }
+    }
+
+    @Transactional
+    public void taskFinished(Task task) {
+        running.remove(task);
+        entityManager.persist(new TaskHistoryRecord("task " + task.getName() + " " + task.getSchedule() + " " +
+                " finished " + System.currentTimeMillis()));
+
+        blocked.stream().filter(task1 -> task1.getDependsOn().contains(task)).forEach(task1 -> {
+            if (Collections.disjoint(task1.getDependsOn(), running)) {
+                blocked.remove(task1);
+                runTask(task1);
+            }
+        });
+    }
+
+    public void skeduleTask(Task task) {
+        for (int i = 0; i < queue.size(); i++)
+            if (task.getSchedule() < queue.get(i).getSchedule()) {
+                queue.add(i, task);
+                return;
+            }
+    }
+
+    public boolean isRunning(String name){
+        return running.stream().anyMatch(task -> task.getName() == name);
+    }
+
+    private void checkRun() {
+        while (queue.get(0).getSchedule() <= System.currentTimeMillis()) {
+            runTask(queue.removeFirst());
```

Comment:
What's your estimation of the number of tasks in queue? The possible problem here: we can run thousands of tasks here at the same time. And since we create a new transaction and new thread for each runTask method call, we can exceed tcp connections limit or even get OOM error. Consider some RPS mechanism here. One possible solution: keep all "runnable" task in another queue and run them with ExecutorService to control number of worker threads.